Switch from in-memory Database to Postgres (pgx) with SQL Migrations

popeskul opened the pull request:
This PR introduces significant changes in our database infrastructure. 

**Summary of Changes**
1. **Database Switch**: Replaced the current in-memory database with a more scalable and persistent Postgres (pgx) database.

2. **Database Migrations**: Added SQL migrations to manage and version control our database schema changes.

3. **Makefile, Dockerfile and Docker-Compose Updates**: The aforementioned files have been updated to reflect the changes associated with moving to a Postgres database.

4. **Repository Update**: The repository layer has been adapted to work with the new Postgres database. There are also minor changes in business logic.

**Impact**

This change allows us to have a more robust and scalable database setup and gives us the ability to manage schema changes in a more controlled manner using migrations.

popeskul committed:
chore: add .vscode/, .idea/ and .env to .gitignore file to avoid committing unnecessary files and sensitive information

popeskul committed:
feat(.env.example): add environment variables for database connection configuration and app health check port

popeskul committed:
feat(.golangci.yml): add linters settings for unused, errcheck, deadcode, and gocyclo linters to improve code quality and maintainability

popeskul committed:
feat: add Dockerfile and docker-compose.yml to enable containerization of the app and its dependencies

The Dockerfile defines a multi-stage build process to create a lightweight container image for the app. The first stage builds the app binary and the second stage copies the binary and other required files to the final image. The image is based on the official Golang and Alpine Linux images.

The docker-compose.yml file defines two services: a PostgreSQL database and the app itself. The app service is built using the Dockerfile in the current directory and depends on the database service. The file also defines environment variables and volumes for the services.

popeskul committed:
db-migration-models: added go.mod sum

popeskul committed:
feat(schema): add initial database schema with users and refresh_sessions tables

This commit adds the initial database schema with two tables: users and refresh_sessions. The users table has columns for id, login, email, birthdate, phone, balance, and created_at. The refresh_sessions table has columns for id, user_id, token, and expires_at. The user_id column in the refresh_sessions table is a foreign key that references the id column in the users table. The uuid-ossp extension is also created if it does not exist.

popeskul committed:
feat: add new files and implement handlers for user creation and retrieval

This commit adds new files and implements handlers for user creation and retrieval. The new files include `app.go`, `config.go`, `postgres.go`, `conversion.go`, `token.go`, `user.go`, `handlers.go`, and `handlers_v1.go`. The handlers are implemented in the `v1` package and include `CreateUser` and `GetUserByFilters`. These handlers are used to create and retrieve user data respectively.

feat: add support for creating and getting users in v1 package
feat: add hash package with HashManager interface and Manager implementation
feat: add health package with StartHealthCheckServer function to start a health check server on a specified port
test(hash_test.go): add tests for hash package
docs: add swagger annotations to CreateUser and GetUserByFilters functions in user.go file
refactor(response.go): remove unused UserRequestFilters struct
refactor(response.go): remove Password field from UserResponse struct
refactor(response.go): change Balance field type from float64 to interface{} in UserRequest struct to support different types of balances
refactor(response.go): convert balance from dollars to cents in CreateUser function before returning UserResponse
refactor(response.go): change newErrorResponse function to return an ErrorResponse struct instead of a string message
refactor(response.go): add comments to exported structs and functions in response.go file
refactor(user.go): add comments to exported functions in user

feat(repository): add SessionsStore struct to handle refresh token creation, retrieval and deletion
feat(repository): add methods to SessionsStore to create, retrieve and delete refresh tokens
feat(repository): add DB interface to combine User and Session interfaces
feat(repository): add NewRepository function to create a new Repository instance with User and Session interfaces
test(repository): add tests for NewRepository function to check if it returns a valid Repository instance and if it returns an error when db is nil

feat(user.go): add UserStore struct and methods CreateUser and GetUserByEmail to interact with users table in the database. Add error variables ErrUserNotFound and ErrUserCreate to handle errors in these methods.

feat: add user service with CreateUser and FindUserByEmail methods
feat: add auth service with GetAccessToken method
test: add unit tests for auth and user services

feat(token): add PasetoManager implementation for token creation and verification using PASETO protocol.

feat(token, conv, currency, random): add new packages and functions to support token generation, type conversion, currency conversion, and random string and integer generation.

popeskul committed:
feat(main.go): add support for reading configuration from environment variables and .env file
feat(main.go): add graceful shutdown of the server on SIGTERM and SIGINT signals

popeskul committed:
feat(Makefile): add Makefile with targets for building, running, testing, and linting the application
feat(Makefile): add docker-up and docker-down targets for managing docker containers
feat(Makefile): add migrate-up and migrate-down targets for running database migrations
feat(Makefile): add lint and lint-fast targets for running linters on the codebase
feat(Makefile): add swag target for generating swagger documentation

popeskul committed:
db-migration-models: udpated swagger

nkolosovgo commented on internal/config/init_config.go:
```diff
@@ -0,0 +1,47 @@
+package config
+
+import (
+	"github.com/kelseyhightower/envconfig"
```
Рекомендую использовать библиотеку
https://github.com/spf13/viper

nkolosovgo commented on internal/db/config.go:
```diff
@@ -0,0 +1,18 @@
+
+import "fmt"
+
+type ConfigDB struct {
```
И так пакет db, можно назвать структуру просто Config

nkolosovgo commented on internal/db/postgres/postgres.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/nkolosov/whip-round/internal/db"
+)
+
+func NewPostgresConnection(cfg db.ConfigDB) (*sql.DB, error) {
```
Конфиг тут стоит передавать по указателю, чтобы не копировать постоянно в процессе выполнения

nkolosovgo commented on internal/db/postgres/postgres.go:
```diff
@@ -0,0 +1,23 @@
+		return nil, err
+	}
+
+	store.SetMaxIdleConns(300)
```
Эти параметры лучше всего тоже задавать через конфигурацию, а дефолтные значения - в константах

nkolosovgo commented on internal/services/auth.go:
```diff
@@ -0,0 +1,42 @@
+	}
+}
+
+func (s *AuthService) GetAccessToken(ctx context.Context, email string, password string) (string, string, error) {
```
А что во втором параметре? Во всех ответах пустая строка.

Тут не хватает документации к коду, чтобы было более очевидно

nkolosovgo commented on internal/utils/currency/currency.go:
```diff
@@ -0,0 +1,13 @@
+
+// ConvertCentsToDollars converts balance from cents to dollars
+func ConvertCentsToDollars(cents int64) float64 {
+	return float64(cents) / 100
```
Для таких целей рекомендую использовать библиотеку
https://github.com/shopspring/decimal

nkolosovgo commented on schema/000001_init.up.sql:
```diff
@@ -0,0 +1,21 @@
+    user_id    UUID         NOT NULL REFERENCES users (id),
+    token      VARCHAR(255) NOT NULL,
+    expires_at TIMESTAMP    NOT NULL
+);
```
А где индексы?

popeskul committed:
refactor(app.go, config.go, postgres.go): rename ConfigDB to Config to improve semantics and consistency across the codebase

popeskul committed:
fix(app.go): pass a pointer to db.Config to NewPostgresConnection function to avoid copying the struct
fix(postgres.go): add a check for nil pointer to db.Config to avoid panics and return an error instead
feat(config.go): add an error variable to be returned when db.Config is nil to improve error handling

popeskul committed:
feat(config): add support for configuring maximum number of idle and open connections to the database
feat(db): set maximum number of idle and open connections to the database based on the configuration
fix(env.example): correct typo in DB_MAX_OPEN_CONNS variable name and add DB_MAX_IDLE_CONNS variable to the example file
refactor(main.go): update initConfig function to read and parse new DB_MAX_IDLE_CONNS and DB_MAX_OPEN_CONNS variables from environment

popeskul committed:
feat(config.go): add support for environment variables to configure server, database, token and hash salt settings.

popeskul committed:
db-migration-models: added modules

popeskul committed:
refactor(main.go): remove initConfig function and move its functionality to config.New() and loadEnv() functions
feat(main.go): add support for loading environment variables using viper package to allow for more flexible configuration options

popeskul committed:
db-migration-models: updated Config with viper

popeskul committed:
db-migration-models: optimized imports

popeskul committed:
db-migration-models: deleted unused imports

popeskul committed:
db-migration-models: added github.com/shopspring/decimal for operations with numbers

popeskul committed:
db-migration-models: added undexes to sql

popeskul committed:
db-migration-models: updated mocks and tests for services, utils, handlers and domain

popeskul committed:
db-migration-models: merged from master

popeskul committed:
db-migration-models: added init-db.sh script for db

popeskul committed:
refactor(db): update Config.String() method to use a URL format for the database connection string

The Config.String() method was updated to use a URL format for the database connection string instead of the previous format. This change improves the readability and maintainability of the code.

popeskul committed:
feat(config): add DB_CONNECT_TIMEOUT to postgres configuration to set a timeout for database connection
feat(app): pass DB_CONNECT_TIMEOUT to postgres connection configuration to set a timeout for database connection
feat(config): add ConnectTimeout field to PostgresConfig to store DB_CONNECT_TIMEOUT value

popeskul committed:
db-migration-models: added pgx

popeskul committed:
refactor(postgres.go): replace sql.DB with pgxpool.Pool to use pgx connection pool library and improve performance
feat(postgres.go): add error wrapping to improve error handling and make it more informative

popeskul committed:
feat(mock_hash.go): add generated mock file for HashManager interface to be used in unit tests

popeskul committed:
feat(token): add generated mocks for Manager interface to be used in unit tests

popeskul committed:
db-migration-models: refactored repository with pgx

popeskul committed:
fix(db): decrease DB_MAX_IDLE_CONNS to 60 to reduce idle connections and improve performance
refactor(app): add context to NewPostgresConnection function to be able to cancel connection attempts and prevent goroutine leaks

popeskul committed:
chore(.gitignore): add whip-round executable to ignore list

popeskul committed:
feat(app.go, pool.go): add support for graceful shutdown of database connection pool to prevent resource leaks and improve reliability

popeskul committed:
refactor(app.go): rename NewPostgresConnection to NewPool to improve semantics
refactor(pool.go): update function name to match new name in app.go
